Public/Src/Cache/ContentStore/Distributed/Redis/RedisDatabaseAdapter.cs

```diff
@@ -286,16 +293,81 @@ public async Task<BoolResult> ExecuteBatchOperationAsync(Context context, IRedis
+        }
+
+        /// <summary>
+        /// This method "handles" redis connectivity errors and reset the connection multiplexer if there are no successful operations happening.
+        /// </summary>
+        private void HandleRedisExceptionAndResetMultiplexerIfNeeded(Context context, Exception exception)
+        {
+            if (IsRedisConnectionException(exception))
+            {
+                // Using double-checked locking approach to reset the connection multiplexer only once.
+                if (Interlocked.Increment(ref _connectionErrorCount) >= _redisConnectionErrorLimit)
+                {
+                    lock (_resetConnectionsLock)
+                    {
+                        // The second read of _connectionErrorCount is a non-interlocked read, but it should be fine because it is happening under the lock.
+                        if (_connectionErrorCount >= _redisConnectionErrorLimit)
+                        {
+                            // This means that there is no successful operations happening, and all the errors that we're seeing are redis connectivity issues.
+                            // This is, effectively, a work-around for the issue in StackExchange.Redis library (https://github.com/StackExchange/StackExchange.Redis/issues/559).
+                            context.Warning($"Reset redis connection due to connectivity issues. ConnectionErrorCount={_connectionErrorCount}, RedisConnectionErrorLimit={_redisConnectionErrorLimit}.");
+                            Interlocked.Exchange(ref _connectionErrorCount, 0);
+                            _databaseFactory.ResetConnectionMultiplexer();
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// The call back that is called by the redis retry strategy.
+        /// </summary>
+        /// <remarks>
+        /// There are two very similar methods: <see cref="OnRedisException"/> and <see cref="HandleRedisExceptionAndResetMultiplexerIfNeeded"/>. The first one (the current method) is called
+        /// by the retry strategy to track and increment the failure count for all the retries not only for the final ones.
+        /// The later is called manually after each failed operation and that method takes a context that allows us to trace and actually may reset the connection mutliplexer
+        /// due to a number of redis connection exceptions.
+        /// </remarks>
+        private void OnRedisException(Exception exception)
+        {
+            // Lets be very pessimistic here: reset the connectivity errors every time when the operation succeed or any other exception
+            // but redis connection exception occurs.
+            if (IsRedisConnectionException(exception))
+            {
+                Interlocked.Increment(ref _connectionErrorCount);
+            }
+            else
+            {
+                Interlocked.Exchange(ref _connectionErrorCount, 0);

```

Comment:
>Interlocked.Exchange(ref _connectionErrorCount, 0); [](start = 16, length = 51)

Question: why are we resetting the error count to 0 when an exception isn't a connection exception? isn't it possible for them to be interleaved with other Redis exceptions? in such a case, we could potentially never reach the required limit

Reply:
Intuitively, I agree with Julian here: not clear why another exception should re-assure us about the one we care about. But I don't think my thinking is ironclad and could be easily swayed.

---
In reply to: [347636933](https://github.com/microsoft/BuildXL/pull/1229#discussion_r347636933) [](ancestors = 347636933)

Reply:
I wanted to be safe here. My idea is that if the error is sticky, then there is no other errors happening besides this one, so it is ok to reset the multiplexer when 10 (or even 5) connectivity issues happened.

If another exception does not reset the count, then the overall connection reset counter should be higher.

On the other hand, successful operation should definitely reset it anyways, so maybe reset is just fine after lets say 20 connectivity issues even if another issues were happening in the process.

---
In reply to: [347642715](https://github.com/microsoft/BuildXL/pull/1229#discussion_r347642715) [](ancestors = 347642715,347636933)

Reply:
stopped resetting the error count.

---
In reply to: [347660296](https://github.com/microsoft/BuildXL/pull/1229#discussion_r347660296) [](ancestors = 347660296,347642715,347636933)